File: 64-minimum-path-sum/minimum-path-sum.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var minPathSum = function (grid) {
    // bottom-up
    const m = grid.length;
    const n = grid[0].length;
    let dp = new Array(n).fill(0);
    dp[0] = grid[0][0];

    for (let i = 0; i < m; i++) {
        let preRow = dp.concat([]);
        for (let j = 0; j < n; j++) {
            let res = Infinity;
            if (i + j === 0) {
                continue;
            }
            if (i > 0) {
                res = Math.min(res, preRow[j]);
            }
            if (j > 0) {
                res = Math.min(res, preRow[j - 1]);
            }
            preRow[j] = res + grid[i][j];
        }
        dp = preRow.concat([]);
    }
    return dp[n - 1];
};

File: 64-minimum-path-sum/minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int[][] minSum = new int[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 && j == 0)
                    minSum[i][j] = grid[i][j];
                else if (j == 0 && i != 0)
                    minSum[i][j] = minSum[i - 1][j] + grid[i][j];
                else if (i == 0 && j != 0)
                    minSum[i][j] = minSum[i][j - 1] + grid[i][j];
                else
                    minSum[i][j] +=  grid[i][j] + Math.min(minSum[i][j - 1], minSum[i - 1][j]);
            }
        }
        return minSum[m - 1][n - 1];
    }
}